### PR TITLE
Bug 1929160: Add NetworkManager reload to resolv-prepender

### PR DESCRIPTION
**Pull request by cybertron**
If resolv.conf gets populated after NetworkManager has attempted to
do a reverse lookup on the node IP to get the hostname, we can end
up with a hostname of localhost because the lookup has no DNS servers
to use. To avoid this possibility, reload NetworkManager after we
populate resolv.conf, which should force it to redo the lookup.

**- Description for the changelog**
Reload NetworkManager after populating resolv.conf to ensure we correctly populate the node hostname.

### CODE REVIEW

**Comment by cybertron** [comment]
/hold

I'm hoping we can get this tested in an environment where the race condition has reproduced pretty consistently.

**Comment by cybertron** [comment]
/retest
/hold cancel

We had a successful test in the failing environment, so I think we're good to merge this.

**Review by mandre — COMMENTED** [review]
/lgtm

**Review by JAORMX — APPROVED** [review]
/retest

**Comment by cybertron** [comment]
/test e2e-vsphere

The vsphere job looks pretty red lately. Might need to check in with that team.

**Comment by cybertron** [comment]
/test e2e-vsphere

Couldn't acquire lease.

**Comment by yuvalk** [comment]
cluster created with this PR still exhibit the issue described in the BZ.
ie one of the masters still shows with 'localhost'
https://bugzilla.redhat.com/show_bug.cgi?id=1929160#c21

**Comment by cybertron** [comment]
/hold

It seems this alone wasn't enough to fix the problem. We're waiting on a NetworkManager change and will revisit this.

**Comment by kikisdeliveryservice** [comment]
@cybertron do you still want this open?

**Comment by cybertron** [comment]
Yes, I think we eventually need a patch like this. However, we need https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/832/diffs?commit_id=43fc98498968ef64edb270c5488c04ad94bdd232 to show up in RHCOS before it will work as intended. I'm not seeing it yet in my local deployments so I don't think this can proceed.

/hold

**Comment by cybertron** [comment]
/test e2e-ovirt
/test e2e-vsphere

I've updated this to use the new nmcli reload command that forces NM to redo the DNS lookup. Hopefully this will fix the problem once and for all.

I'm going to leave the hold because it would be nice to test this in an environment where the problem reproduced.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml
Comment (mandre) [inline templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:78]: Strangely, `nmcli` version 1.30.6-1.fc34 doesn't show the reload command as a valid command:
```
nmcli general -h     
Usage: nmcli general { COMMAND | help }

COMMAND := { status | hostname | permissions | logging }

  status

  hostname [<hostname>]

  permissions

  logging [level <log level>] [domains <log domains>]
```

However, it's rightfully documented in the man page:
```
nmcli general {status | hostname | permissions | logging | reload} [ARGUMENTS...]
```

The man page also says that the `reload dns-rc` command "Update DNS configuration, which usually involves writing /etc/resolv.conf anew". Isn't it a problem for us? Won't it overwrite our changes to `/etc/resolv.conf`?

Comment (cybertron) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:78]: Yeah, I actually delayed updating this patch because it wasn't showing up in the cli, but at some point I realized it was there anyway. :-)

The reason I used dns-rc is because https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/832 is the patch in response to this bug and it mentions using SIGUSR1 to notify NM that the DNS config had changed. In the docs, dns-rc corresponds to SIGUSR1. In my local tests it seemed to work fine, but I'm waiting on the bug reporter to try it in an environment where the bug reproduces. If it doesn't work then I may just do a general reload without parameters to reload everything.

**Comment by cybertron** [comment]
/hold cancel

I'm still not aware of anyone having tested this in an environment where the problem reproduces, but based on the investigation of the problem so far there's a reasonable expectation that this will fix it and there's no point holding out a probably-good patch while we wait for pre-verification. I'm going to propose that we merge this and if it ends up not fixing the problem we can revisit.

**Comment by yuvalk** [comment]
> /hold cancel
> 
> I'm still not aware of anyone having tested this in an environment where the problem reproduces, but based on the investigation of the problem so far there's a reasonable expectation that this will fix it and there's no point holding out a probably-good patch while we wait for pre-verification. I'm going to propose that we merge this and if it ends up not fixing the problem we can revisit.

I own a lab where issue can be reproduced, but currently we are running with this WA - 
https://github.com/openshift-kni/cnf-features-deploy/tree/master/feature-configs/deploy/localhost-workaround
I didnt have time to first try without it, make sure is reproducible then test with this PR ...

**Comment by LorbusChris** [comment]
We get reports with hostname set to `localhost` rather frequently for OKD. Latest example: https://github.com/openshift/okd/discussions/1039

**Comment by cybertron** [comment]
Yes, my reasoning from https://github.com/openshift/machine-config-operator/pull/2488#issuecomment-983828260 still holds. I'd like to get it in and verify the bug after. It will be easier once it's in real builds and doesn't require pulling down a PR.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml
Comment (mandre) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:78]: For reference, here is a excerpt of what `NetworkManager` man page says about `SIGUSR1`:
```
When NetworkManager is not managing DNS, the signal forces a restart of operations that depend on the DNS
configuration (like the resolution of the system hostname via reverse DNS, or the resolution of WireGuard peers);
therefore, it can be used to tell NetworkManager that the content of resolv.conf was changed externally. In the future,
further actions may be added.
```

So it sounds like something we should do, since we set `rc-manager=unmanaged` to tell NM not to write to `/etc/resolv.conf`.

**Review by mandre — COMMENTED** [review]
/lgtm

**Pull request merged**
